src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -1247,20 +1249,25 @@ BlankNode Visitor::visitTypesafe(Parser::BlankNodeContext* ctx) {
 }
 
 // ____________________________________________________________________________________
-template <typename Out, typename FirstContext, typename... Context>
-Out Visitor::visitAlternative(FirstContext ctx, Context... ctxs) {
+template <typename Out, bool isRecursive, typename FirstContext,
+          typename... Context>
+Out Visitor::visitAlternative(FirstContext* ctx, Context*... ctxs) {
+  if constexpr (!isRecursive) {
+    int validContexts =
+        (static_cast<bool>(ctx) + ... + static_cast<bool>(ctxs));
+    AD_CHECK(validContexts == 1);
+  }
   if (ctx) {
     return {visitTypesafe(ctx)};
   } else {
-    return visitAlternative<Out>(ctxs...);
+    if constexpr (sizeof...(Context) != 0) {
+      return visitAlternative<Out, true>(ctxs...);
+    } else {
+      AD_FAIL()  // Should be unreachable.
+    }
```

Comment:
Ok, but if the recursion is getting clumsy, then we can still do the folding (you know the recursive case, now you can learn something.

```
template<typename Out, typename... Contexts>
Out visitAlternative(Contexts*... ctxs) {
  AD_CHECK(1 == (... + static_cast<bool>(ctxs)));
  std::optional<Out> opt;  // optional is assignable.
  (..., visitIf(&opt, ctxs)); // calls (visitIf(&opt, firstCtx), visitIf(&opt, secondCtx));
  return opt.value();
}
```
This is the shortest way. It will still check the nullptrs after the matching pointer, but we can afford this. If this wasn't desired, you can use a lambda that returns a bool and sets the optional instead of `visitIf` and then use `||` in the fold expression instead of `,`, but I don't think there's much to gain here:)
```

Reply:
This would necessitate that `visitIf` does some sort of conversion on the return of `visitTypesafe`. At the moment this is `SparqlValues` to `Values`. Later it will even be the explicit constructor of `Literal`.
With that comes that we'd also have to explicitly specify the type we want in `visitIf`. I think I'd rather leave it in the state it currently is.

Reply:
Yes that is fine for now, maybe I'll make a draft for that.